lib/rules/no-misleading-character-class.js

```diff
@@ -181,6 +231,50 @@ module.exports = {
         const sourceCode = context.sourceCode;
         const parser = new RegExpParser();
 
+        /**
+         * Generates a granular loc for context.report.
+         * @param {Character[]} chars Individual characters being reported on.
+         * @param {Node} node Parent string node to report within.
+         * @returns {Object} Granular loc for context.report.
+         * @see https://github.com/eslint/eslint/pull/17515
+         */
+        function generateReportLocation(chars, node) {
+
+            // We only attempt to use a granular location for literals without dynamic expressions
+            if (node.type !== "Literal" && (node.type !== "TemplateLiteral" || node.expressions.length)) {
+                return node.loc;
+            }
+
+            const rawText = sourceCode.getText(node);
+
+            // Factoring in backslash-escaped characters, bail if the first character doesn't match real text
+            const first = chars[0];
+            const reportStartOffset = countDoubleEscapedBackslashes(rawText.slice(0, first.start));
+
+            if (!rawText.slice(first.start, first.end + 1 + reportStartOffset).includes(first.raw)) {
+                return node.loc;
+            }
+
+            // Factoring in backslash-escaped characters, also bail if the last character doesn't match real text
+            const last = chars[chars.length - 1];
+            const reportEndOffset = countDoubleEscapedBackslashes(rawText.slice(first.start, last.end));
+
+            if (!rawText.slice(last.start, last.end + last.parent.raw.length + reportEndOffset).includes(last.raw)) {
+                return node.loc;
+            }
+
+            return {
+                end: {
+                    line: node.loc.end.line,
+                    column: node.loc.start.column + last.end + 1 + reportEndOffset + reportStartOffset
+                },
+                start: {
+                    line: node.loc.start.line,
+                    column: node.loc.start.column + first.start + 1 + reportStartOffset
+                }
+            };
```

Comment:
This can still find incorrect locations:

![image](https://github.com/eslint/eslint/assets/44349756/860d5fc6-f545-4493-a59d-4e2c6c261c91)

We should limit reporting granular locations to regex literals, and only string/template literals that are exactly the same as their value.